### ISSUE TITLE
Add Check Support for Java 21 Pattern Matching for Switch Syntax: New Check WhenShouldBeUsed

### ISSUE THREAD

**Issue by mahfouz72** [checkstyle/checkstyle#14985]
child of #14961: 

pattern matching for switch, allows more concise and readable code when dealing with various object types. This feature includes the `when` keyword to add conditions to case labels directly, improving readability.

I propose to Introduce a new check to enforce the use of the `when` keyword in switch case patterns to replace single if statements, in the case body, enhancing code readability and consistency.

---
**Config**
```XML
$ cat config.xml 
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
        "-//Checkstyle//DTD Checkstyle Configuration 1.3//EN"
        "https://checkstyle.org/dtds/configuration_1_3.dtd">

<module name="Checker">
    <module name="TreeWalker">
        <module name="WhenShouldBeUsed"/>
    </module>
</module>
```
---
**Example**

```Java
void test(Object response) {
    switch (response) {
        case String s -> {
            if(s.length() > 80) {     // violation
                System.out.println("This string is too long");
            }
         }
        default -> System.out.println("Unknown type");
    }
     switch (response) {
        case String s when s.length() > 80 -> {  // ok
                System.out.println("This string is too long");
         }
        default -> System.out.println("Unknown type");
    }
}
```
**Expected Output**
```

$ java -jar checkstyle-10.19.0-all.jar -c config.xml Test.java 
Starting audit...
[ERROR] D:\Test.java:4:12: when can be used instead of single if statement. [WhenShouldBeUsed]
Audit done.
Checkstyle ends with 1 errors.
```

**Comment by rnveach** [comment]
Examples don't show but what if there is an else to the if or things besides the if statement? Are we going to print no violation?

**Comment by mahfouz72** [comment]
Yes, no violation. IMO if we have more than one condition to check in the case body then a series of  `if elif else` will be more readable than using `when` However I don't have a strong opinion on this so I am not sure. what do you think?
please note that the sonarlint rule works when there is a single if inside the pattern body

```Java
        switch (response) {
            case String s -> {
                if (s.length() > 100) {
                    System.out.println("This string is too long");
                } else if (s.length() > 10) {
                    System.out.println("This string is long");
                } else if (s.length() > 5) {
                    System.out.println("This string is meduim");
                } else {
                    System.out.println("This string is short");
                }
            }
            default -> System.out.println("Unknown type");
        }
        // after fix
        switch (response) {
            case String s when s.length() > 100 -> {
                System.out.println("This string is too long");
            }
            case String s when s.length() > 10 -> {
                System.out.println("This string is long");
            }
            case String s when s.length() > 5 -> System.out.println("This string is meduim");
            case String s -> System.out.println("This string is short");
            default -> System.out.println("Unknown type");
        }
```

**Comment by rnveach** [comment]
> if elif else will be more readable than using when However I don't have a strong opinion on this so I am not sure. what do you think?

I say what till after this check is implemented before we think about this. I would just like issue descriptions for new checks to be more robust so there is no surprises in PR. We should be communicating what we will validate and what we won't validate.

**Comment by nrmancuso** [comment]
I am approving for single `if` statement only. Let's keep the first implementation simple.

**Comment by rnveach** [comment]
Identified at https://github.com/checkstyle/checkstyle/pull/15113#discussion_r1663471696 ,

The bytecode when using `when` is nearly identical to just using an nested if statement inside the case.
